qiskit/algorithms/optimizers/fraxis.py

```diff
@@ -0,0 +1,171 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Free-Axis Selection (Fraxis) algorithm."""
+
+from typing import Optional, Tuple
+
+import numpy as np
+from scipy.optimize import OptimizeResult
+
+from qiskit.quantum_info import OneQubitEulerDecomposer, Pauli
+from qiskit.utils import algorithm_globals
+
+from .scipy_optimizer import SciPyOptimizer
+
+
+class Fraxis(SciPyOptimizer):
+    """
+    Free-Axis Selection (Fraxis) algorithm [1].
+
+    More precisely, this class implements π-Fraxis algorithm in Algorithm 1 of [1].
+
+    .. note::
+
+        This optimizer only works with U gates as parameterized gates etc.
+
+    References:
+      [1] "Optimizing Parameterized Quantum Circuits with Free-Axis Selection,"
+          HC. Watanabe, R. Raymond, Y. Ohnishi, E. Kaminishi, M. Sugawara
+          `arXiv:2104.14875 <https://arxiv.org/abs/2104.14875>`__
+    """
+
+    _OPTIONS = ["maxiter", "disp", "initialize"]
+
+    # pylint: disable=unused-argument
+    def __init__(
+        self,
+        maxiter: Optional[int] = None,
+        disp: bool = False,
+        initialize: bool = True,
+        options: Optional[dict] = None,
+        **kwargs,
+    ) -> None:
+        """
+        Args:
+            maxiter: Maximum number of iterations to perform.
+                Will default to N*10, where N is the number of U gates
```

Comment:
`x0.size // 3` means the number of U gate. I will mention it.

Reply:
done

Reply:
I figured thats what it meant.

>Basically I guess it requires all 3 U gate angles to be individually parameterized. Maybe some more summary/overview in the class docstring would help for points such as this. A reference is good for people wanting more detail but it would be nice if there was enough explanation for me to understand/use it without having to resort to reading the ref.

I still think some summary in the docstring would be helpful. Otherwise its pretty slim. I imagine each U gate has to have each parameter parameterized right. Two U gates where three of the parameters were fixed values and other three parameterized would meet the need of parameter len divisible by 3, but I assume this is not a valid circuit as such, even though its U gates.

Reply:
Yes, each U gate should have each parameter parameterized (i.e., U(x, y, z), x, y, and z are parameters) or each parameter fixed (i.e., U(1, 2, 3)). Fixed U gates are OK because `x0` does not include any information of the gates.
I think "number of U gates" is ambiguous because it's not clear whether fixed U gates are included or not.
So, I'm thinking of using "number of parameters" instead of "number of U gates".
For example, `x0.size // 3 * 10` -> `x0.size * 3 (or 4)`.
Does it make sense?

Reply:
I guess one could say

Will default to N*10,  where N is len(x0)/3 and should correspond to the number of fully parametrized U gates in the circuit.

As far as I can see the optimizer will run until maxiters is reached right. There is no early termination. Using a multiply of 4 (I assume that's what (or 4) in your comment above meant) will mean more evaluations by default, 3 slightly less. Maybe a simple multiplier of 3 or 4 works just as well, but since it seems it will go for that many iters clearly the latter means more by default.

Reply:
I see. Currently `maxiter` or callback can terminate the loop. Because I updated the callback to follow scipy manner, it terminates the loop if the callback return True.

It might be good to have an early termination depending on a tolerance of termination. I will think of it.

Reply:
As for the default value of `maxiter`, I want to simplify the code and docs. So, I will try to rewrite it using "number of parameters".

Reply:
Maxiter is like maxiter in SPSA, normally it is the number of iterations since there is no early termination like people see with COBYLA or SLQSP etc that have tol and/or ftol. It does lead to people asking about the behavior where it does not seem to stop. Yes, someone could add an early terminate based on information passed back in the callback, but its not the default. By default SPSA has 100 as maxiter (plus it will do a few more evals for calibration). By default this optimizer will depend on the number of parameters /parametrized U gates and run until that limit is reached. If you picked a suitably small xtol, where its consider converged, does that normally happen before the default maxiter? Of course with a random initial point it maybe highly dependent on that. Just wondering if it would be helpful to have an actual default value for xtol - None could still be permitted, but maybe by default earlier termination might happen?

Will you be adding tests for callback, xtol etc

Reply:
I added unit tests of maxiter, xtol, and callback.